daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala

```diff
@@ -11,10 +14,27 @@ sealed abstract class MatchingStringModule {
   def fromString(s: String): Either[String, T]
 
   @throws[IllegalArgumentException]
-  def assertFromString(s: String): T =
+  final def assertFromString(s: String): T =
     fromString(s).fold(e => throw new IllegalArgumentException(e), identity)
 
-  def unapply(x: T): Some[String] = Some(x)
+  def equalInstance: Equal[T]
+
+  // We provide the following array factory instead of a ClassTag
+  // because the latter lets people easily reinterpret any string as a T.
+  // See
+  //  * https://github.com/digital-asset/daml/pull/983#discussion_r282513324
+  //  * https://github.com/scala/bug/issues/9565
+  val Array: ArrayFactory[T]
+
+}
+
+sealed abstract class ArrayFactory[T](implicit classTag: ClassTag[T]) {
+
+  def apply(xs: T*): Array[T] = xs.toArray
+
+  def ofDim(n: Int): Array[T] = Array.ofDim(n)
```

Comment:
Exactly what I had in mind.

There is an implication you should be aware of: `null` is generally rejected as an inhabitant of `Name`, in other words `null: Name` fails at type-checking, not wartremover. However, IIUC `ofDim` effectively backdoors a `null` inhabitant. This isn't really enough of a bother to avoid this, necessarily (you can also get a `null` inhabitant by getting the 0th item of an empty java ArrayList), but worth keeping in mind.

We might also want to provide a `CanBuildFrom[Array[_], T, Array[T]]` instance, but see my other comment.